src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+                }
+
+                auto processRename = [item, originalPath, base, this](PathTuple &path) {
+                    auto adjustedOriginalPath = _discoveryData->adjustRenamedPath(originalPath);
+                    _discoveryData->_renamedItems.insert(originalPath, path._target);
+                    item->_modtime = base._modtime;
+                    item->_inode = base._inode;
+                    item->_instruction = CSYNC_INSTRUCTION_RENAME;
+                    item->_direction = SyncFileItem::Up;
+                    item->_renameTarget = path._target;
+                    item->_file = adjustedOriginalPath;
+                    item->_originalFile = originalPath;
+                    item->_fileId = base._fileId;
+                    item->_remotePerm = base._remotePerm;
+                    item->_etag = base._etag;
+                    item->_type = base._type;
+                    path._original = originalPath;
+                    path._server = adjustedOriginalPath;
+                    qCInfo(lcDisco) << "Rename detected (up) " << item->_file << " -> " << item->_renameTarget;
+                };
+                if (isMove && !oldEtag.isEmpty()) {
+                    recurseQueryServer = oldEtag == base._etag ? ParentNotChanged : NormalQuery;
+                    processRename(path);
+                } else if (isMove) {
+                    // We must query the server to know if the etag has not changed
+                    _pendingAsyncJobs++;
+                    QString serverOriginalPath = originalPath;
+                    if (localEntry.isVirtualFile)
+                        serverOriginalPath.chop(_discoveryData->_syncOptions._virtualFileSuffix.size());
+                    auto job = new RequestEtagJob(_discoveryData->_account, serverOriginalPath, this);
+                    connect(job, &RequestEtagJob::finishedWithResult, this, [=](const Result<QString> &etag) mutable {
+                        if (!etag || (*etag != base._etag && !item->isDirectory()) || _discoveryData->_renamedItems.contains(originalPath)) {
+                            qCInfo(lcDisco) << "Can't rename because the etag has changed or the directory is gone" << originalPath;
+                            // Can't be a rename, leave it as a new.
+                            postProcessLocalNew();
+                        } else {
+                            // In case the deleted item was discovered in parallel
+                            auto it = _discoveryData->_deletedItem.find(originalPath);
+                            if (it != _discoveryData->_deletedItem.end()) {
+                                ASSERT((*it)->_instruction == CSYNC_INSTRUCTION_REMOVE);
+                                (*it)->_instruction = CSYNC_INSTRUCTION_NONE;
+                            }
+                            delete _discoveryData->_queuedDeletedDirectories.take(originalPath);
```

Comment:
similar check-async-check pattern; possibly make the "check" part a function again?